beancount_import/source/amazon_invoice.py

```diff
@@ -15,6 +17,256 @@
+
+    # Payment Table & Credit Card Transactions
+    grand_total=r'\n\s*Grand Total:\s+(.*)\n'
+    credit_card_transactions='Credit Card transactions'
+    credit_card_last_digits=r'^([^:]+) ending in ([0-9]+):\s+([^:]+):$'
+    payment_type=[
+        # only first matching regex is used!
+        r'\n\s*([^\s|][^|\n]*[^|\s])\s+\|\s+Last (?:4 )?digits:\s+([0-9]{4})\n',
+        r'\n\s*(.+)\s+ending in\s+([0-9]{4})\n'
+        ]
+    payment_information='^Payment information$'
+
+    # regular orders only
+    shipment_shipped_pattern='^Shipped on ([^\\n]+)$'
+    shipment_nonshipped_headers=[
+        'Service completed',
+        'Preparing for Shipment',
+        'Not Yet Shipped',
+        'Shipping now'
+        ]
+    shipment_quantity=r'^\s*(?:(?P<quantity>[0-9]+)|(?P<weight1>[0-9.]+\s+(?:lb|kg))|(?:(?P<quantityIgnore>[0-9.]+) [(](?P<weight2>[^)]+)[)]))\s+of:'
+    shipment_of='of:'
+    shipment_sales_tax='Sales Tax:'
+    shipment_total='Total for This Shipment:'
+    shipment_seller_profile=' (seller profile)'
+    shipment_sold_by=r'(?P<description>.*)\n\s*(?:Sold|Provided) by:? (?P<sold_by>[^\n]+)'
+    shipment_condition=r'\n.*\n\s*Condition: (?P<condition>[^\n]+)'
+    regular_total_order='Grand Total:'
+    regular_estimated_tax = 'Estimated tax to be collected:'
+    regular_order_placed=r'(?:Subscribe and Save )?Order Placed:\s+([^\s]+ \d+, \d{4})'
+    regular_order_id=r'.*Order ([0-9\-]+)'
+    gift_card='Gift Cards' # not confirmed yet!
+    gift_card_to=r'^(?P<type>Gift Card)[\w\s-]*:\s*(?P<sent_to>[\w@._-]*)$' # guess, not confirmed yet!
+    gift_card_amazon_account=r'^[\w\s-]*(?P<type>Amazon-Account)[\w\s-]*(?P<sent_to>charged up)[\w\s-]*$' # guess, not confirmed yet!
+
+    # digital orders only
+    digital_order='Digital Order: (.*)'
+    digital_order_cancelled='Order Canceled'
+    digital_by='By'
+    digital_sold_by=r'Sold\s+By'
+    digital_tax_collected='Tax Collected:'
+    digital_total_order='Total for this Order:'
+    digital_order_id='^Amazon.com\\s+order number:\\s+(D[0-9-]+)$'
+    digital_payment_information='Payment Information'
+
+    @staticmethod
+    def parse_amount(amount, assumed_currency=None) -> Amount:
+        return parse_amount(amount, assumed_currency=assumed_currency)
+
+    @staticmethod
+    def parse_date(date_str) -> datetime.date:
+        return dateutil.parser.parse(date_str).date()
+
+
+class Locale_de_DE(Locale_Data):
+    """Language and region specific settings for parsing amazon.de invoices
+    """
+    LOCALE='de_DE'
+    tax_included_in_price=True  # no separate tax transactions
+    payee='Amazon.de'
+
+    # common fields regular and digital orders
+    items_ordered='Bestellte Artikel|Erhalten|Versendet|Amazon-Konto erfolgreich aufgeladen' # Erhalten|Versendet for gift cards
+    price='Preis|Betrag'
+    currency='EUR'
+    items_subtotal='Zwischensumme:'
+    total_before_tax='Summe ohne MwSt.:'
+    # most of translations still missing ...
```

Comment:
unfortunately, there is no way to get all these translations (if they exist at all). I parsed all my amazon transactions from the last years and added every case that occurred. Probably there will be minor PRs in the future adding additional translations...

Reply:
So what happens if a missing translated string is needed? Presumably an exception because the base class is abstract.

Do you see a way for us to handle this gracefully?
We'd ideally notify the user of the fact that we cannot parse their invoice, not crash, and even suggest they open an issue.

Reply:
exceptions should not happen. The base class is abstract but all attributes are (and have to be!) set in the derived class. Only alternatives may be missing which will not lead to exceptions, but just missed cases.